### PR TITLE
Add the RetryResult class

### LINKED ISSUE #245: Add RetryResult

### PR DESCRIPTION
**Pull request by sleberknight**
Fixes #245

### CODE REVIEW

File: src/main/java/org/kiwiproject/retry/RetryResult.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+     * Create new instance.
+     *
+     * @param numAttemptsMade the number of attempts (must be less than or equal to max attempts)
+     * @param maxAttempts     the maximum number of attempts
+     * @param object          the result, or {@code null}
+     * @param errors          a non-null list containing any errors that occurred on attempts, or empty list if no
+     *                        errors occurred
+     */
+    public RetryResult(int numAttemptsMade, int maxAttempts, T object, List<Exception> errors) {
+        this.resultUuid = UUIDs.randomUUIDString();
```
Comment (chrisrohr) [inline src/main/java/org/kiwiproject/retry/RetryResult.java:50]: Maybe we should provide a check to ensure that `numAttemptsMade` is less than or equal to `maxAttempts`

File: src/main/java/org/kiwiproject/retry/RetryResult.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        this.numAttemptsMade = numAttemptsMade;
+        this.maxAttempts = maxAttempts;
+        this.object = object;
+        this.errors = errors;
```
Comment (chrisrohr) [inline src/main/java/org/kiwiproject/retry/RetryResult.java:54]: We should probably protect the list from being changed by making a copy.

```suggestion
        this.errors = List.copyOf(errors);
```

File: src/main/java/org/kiwiproject/retry/RetryResult.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    /**
+     * Return a set containing the unique error types in this result. There is NO guarantee whatsoever about the
+     * order of the unique error types, so don't make any assumptions. That's why the return type is {@link Set}.
+     *
+     * @return set of unique error types (class names)
+     */
+    public Set<String> getUniqueErrorTypes() {
+        return errors.stream()
+                .map(e -> e.getClass().getName())
+                .collect(Collectors.toSet());
```
Comment (chrisrohr) [inline src/main/java/org/kiwiproject/retry/RetryResult.java:169]: Statically import `toSet()`

File: src/main/java/org/kiwiproject/retry/RetryResult.java (hunk 2 above)
Comment (chrisrohr) [inline src/main/java/org/kiwiproject/retry/RetryResult.java:54]: What happens if a null `errors` is passed in?  We should support that case.

File: src/test/java/org/kiwiproject/retry/RetryResultTest.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        @Test
+        void shouldHaveMaxAttempts() {
+            assertThat(result.getMaxAttempts()).isEqualTo(5);
+        }
+
+        @Test
+        void shouldHaveNoErrors() {
+            assertThat(result.hasAnyErrors()).isFalse();
+            assertThat(result.getNumErrors()).isEqualTo(0);
```
Comment (chrisrohr) [inline src/test/java/org/kiwiproject/retry/RetryResultTest.java:66]:
```suggestion
            assertThat(result.getNumErrors()).isZero();
```

File: src/test/java/org/kiwiproject/retry/RetryResultTest.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                    new IllegalArgumentException("attempt 5 error")
+            ));
+            assertThat(result.getUniqueErrorTypes()).containsExactlyInAnyOrder(
+                    IllegalAccessException.class.getName(),
+                    IllegalArgumentException.class.getName(),
+                    IllegalStateException.class.getName()
+            );
+        }
+    }
+}
```
Comment (chrisrohr) [inline src/test/java/org/kiwiproject/retry/RetryResultTest.java:216]: Add a newline at the end of the file

File: src/main/java/org/kiwiproject/retry/RetryResult.java (hunk 2 above)
Comment (sleberknight) [reply src/main/java/org/kiwiproject/retry/RetryResult.java:54]: Done

Comment (sleberknight) [reply src/main/java/org/kiwiproject/retry/RetryResult.java:54]: Added new issue #247

**Pull request merged**